File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num % (long) pow(10, 12) == num || num % (long) pow(10, 16) != num);

//Calculating the checksum as per Luhn’s Algorithm

    int sum = 0;
    int digit = 0;
    for (int i = 1; i <= 16; i++)
    {
        if (num % (long) pow(10, i-1) == num)
        {
            break;
        }
        digit = floor(num % (long) pow(10, i) / pow(10, (i-1)));
        if (i % 2 == 0)
        {
            sum = (digit * 2 > 9) ? sum + ((digit * 2) % 10 + floor (digit * 2 / 10)) : sum + digit * 2;
        }
        else
        {
            sum += digit;
        }
    }

//Checking the card number

    if (sum % 10 == 0)
    {
        switch(digit)
        {
            case 3:
                printf("AMEX\n");
                break;
            case 4:
                printf("VISA\n");
                break;
            case 5:
                printf("MASTERCARD\n");
                break;
            default:
                printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
from cs50 import get_string
import re

// Input card number
card = str()
while not card.isdigit():
    card = get_string("Card: ")

// Checksum card number
sum = 0
card = card[::-1]
for i in range(len(card)):
    if i % 2 == 0:
        sum += int(card[i])
    else:
        if int(card[i])*2 > 9:
            sum += int(card[i]) * 2 % 10 + 1
        else:
            sum += int(card[i]) * 2

// Check card validity by checksum
card = card[::-1]
if sum % 10 != 0:
    print('INVALID')

// Check card type using regular expressions
if re.match(r'3[47]\d{13}', card):
    print('AMEX')
elif re.match(r'5[1-5]\d{14}', card):
    print('MASTERCARD')
elif re.match(r'4\d{12|15}', card):
    print('VISA')
else:
    print('INVALID')

